upload/upload/main.py

```diff
@@ -0,0 +1,202 @@
+  '%(message)s')
+
+fh = logging.FileHandler('upload.log')
+fh.setLevel(logging.INFO)
+fh.setFormatter(fmt)
+
+ch = logging.StreamHandler()
+ch.setLevel(logging.INFO)
+ch.setFormatter(fmt)
+
+log = logging.getLogger('upload')
+log.setLevel(logging.INFO)
+
+logging.basicConfig(
+  handlers=[fh, ch],
+  level=logging.INFO)
+
+USERS = set([
+    'cseed@broadinstitute.org',
+    'dking@broadinstitute.org',
+    'jbloom@broadinstitute.org',
+    'jigold@broadinstitute.org',
+    'cvittal@broadinstitute.org',
+    'tpoterba@broadinstitute.org',
+    'pschultz@broadinstitute.org',
+    'wang@broadinstitute.org'
+])
+
+CLIENT_SECRET = '/upload-oauth2-client-secret/client_secret.json'
+SCOPES = [
+  'https://www.googleapis.com/auth/userinfo.email',
+  'https://www.googleapis.com/auth/userinfo.profile',
+  'https://www.googleapis.com/auth/plus.me'
+]
+
+counter = 0
+def next_id():
+    global counter
+
+    counter = counter + 1
+    return counter
+
+id_item = {}
+items = []
+
+class Item(object):
+    def __init__(self, data):
+        self.id = next_id()
+        
+        items.insert(0, self)
+        id_item[self.id] = self
+
+        self.timestamp = datetime.datetime.now()
+        self.data = data
+
+    def typ(self):
+        return self.data['type']
+
+    def email(self):
+        return self.data['email']
+
+    def config(self):
+        return self.data['config']
+
+    def hail_version(self):
+        return self.config()['hail_version']
+    
+    def contents(self):
+        return self.data['contents']
+
+    def preview(self):
+        s = self.contents()
+        if len(s) > 40:
+            s = s[:40] + " ..."
+        
+        return re.sub('\\s+', ' ', s)
+        
+        return s
+
+app = Flask('upload')
+
+with open('/flask-secret-key/flask-secret-key', 'rb') as f:
+    app.secret_key = f.read()
+
+def get_flow(state=None):
+  flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+    CLIENT_SECRET, scopes=SCOPES, state=state)
+
+  # FIXME _scheme='https'
+  flow.redirect_uri = url_for('oauth2callback', _external=True, _scheme='https')
+  
+  return flow
+
+@app.route('/', methods=['GET'])
+def index():
+    if 'email' not in session:
+        return redirect(url_for('login'))
+
+    email = session['email']
+    
+    # just show 50
+    page_items = items[:50]
+    
+    return render_template('index.html', email=email, page_items=page_items, n_items=len(items))
+
+@app.route('/login', methods=['GET'])
+def login():
+  flow = get_flow()
+  authorization_url, state = flow.authorization_url(
+    access_type='offline',
+    include_granted_scopes='true')
+
+  session['state'] = state
+
+  return redirect(authorization_url)
+
+@app.route('/oauth2callback', methods=['GET'])
+def oauth2callback():
+  authorization_response = request.url
+  # FIXME as with url_for(..., _scheme='https')
+  authorization_response = re.sub('^http://', 'https://', authorization_response)
```

Comment:
this is due to nginx stripping HTTPS, yeah?

Reply:
Yes, related to the _scheme='https' stuff above.